app/models/king.rb

```diff
@@ -1,7 +1,55 @@
 class King < Piece
+  def valid_move?(to_row, to_col)
+    logger.debug "*** KING: #{id} #{type} valid_move?"
+    return false if move_nil?(to_row, to_col)
+    return false if move_out_of_bounds?(to_row, to_col)
+    return false if move_destination_ally?(to_row, to_col)
+    return false unless move_legal?(to_row, to_col)
+    true
+  end
+
   def move_legal?(to_row, to_col)
+    logger.debug "move_legal for #{type}"
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
-    row_diff <= 1 && col_diff <= 1
+    return true if row_diff <= 1 && col_diff <= 1
+
+    legal_castle_move(to_row, to_col)
+  end # end move_legal?
+
+  private
+
+  def legal_castle_move(to_row, to_col)
+    return false unless legal_king_move?(to_row, to_col)
+    legal_rook_move?(to_col)
+  end
+
+  def legal_king_move?(to_row, to_col)
+    return false unless moves.empty?
+    return false unless row == to_row
+    return false unless (col - to_col).abs == 2
+    true
+  end
+
+  def legal_rook_move?(to_col)
+    rook = get_rook?(to_col)
+    return false unless rook
+    return false if move_obstructed?(row, rook.col)
```

Comment:
That is true - a rook can move all over the board and return to its original position.

The Moves table contains all of the moves. The method rook_at queries that table as the check that the rook has not moved (line 49). The rook's position on the board is not used to determine that it has never, ever moved. The method only makes use of the  position to determine if a rook even exists at 0,0 || 0,7 || 7,0 || 7,7. 

I have manually tested this concept with IRB and the web application. I moved the rook at 7,7 to a location [3,7] and moved it back to[ 7,7]. I perform a page refresh between every move to ensure that the move happened. Then I manually queried the table for Rook.find(90).moves.empty?. It returned false - because the rook moved. Next I did Move.delete_all and the query again. It returned true to indicate that the rook had not ever moved.

king_spec.rb contains this test as the last one. A move is explicitly created for the rook. In the previous test, a move is created for the king.) Next move_to! is called for the king. The move cannot occur since the rook has already moved.

In order to castle, the method rook_at does enforce the rule that a castle can only occur when it is the first move for both the king and the rook.